plugins/nlp/manifest.cpp

```diff
@@ -0,0 +1,174 @@
+                }
+            }
+        }
+
+        if (!manifest)
+        {
+            DBGLOG("Failed to initialize nlp manifest");
+            manifest.setown(createPTree());
+        }
+    }
+
+public:
+    Manifest(const Manifest &) = delete;
+    Manifest &operator=(const Manifest &) = delete;
+    ~Manifest()
+    {
+        DBGLOG("MANIFEST:  Destructor");
+    }
+
+    static std::shared_ptr<IManifest> getInstance(ICodeContext *ctx)
+    {
+        std::call_once(initFlag, [ctx]()
+                       { instance = std::shared_ptr<Manifest>(new Manifest(ctx)); });
+        return instance;
+    }
+
+    const char *stripLeadingCharacters(const char *input) const
+    {
+        while (*input == '.' || *input == '/' || *input == '\\')
+            ++input;
+        return input;
+    }
+
+    StringBuffer &makeResourcePath(const char *path, const char *basedir, StringBuffer &respath) const
+    {
+        StringBuffer abspath;
+        makeAbsolutePath(path, basedir, abspath);
+        return makePathUniversal(abspath, respath);
+    }
+
+    bool getResourceData(IPropertyTree *res, size32_t &len, const void *&data) const
+    {
+        if (res->hasProp("@id") && (res->hasProp("@header") || res->hasProp("@compressed")))
+        {
+            int id = res->getPropInt("@id");
+            return (dll->getResource(len, data, res->queryProp("@type"), (unsigned)id) && len > 0);
+        }
+        return false;
+    }
+
+    bool getResourceData(IPropertyTree *res, MemoryBuffer &content) const
+    {
+        size32_t len = 0;
+        const void *data = NULL;
+        if (getResourceData(res, len, data))
+        {
+            if (res->getPropBool("@compressed"))
+                decompressResource(len, data, content);
+            else
+                content.append(len, (const char *)data);
+            return true;
+        }
+        return false;
+    }
+
+    IPropertyTree *queryResourcePT(const char *_path) const
+    {
+        const char *path = stripLeadingCharacters(_path);
+        StringBuffer xpath;
+        if (!manifestDir.length())
+            xpath.setf("Resource[@filename='%s'][1]", path);
+        else
+        {
+            StringBuffer respath;
+            makeResourcePath(path, manifestDir.str(), respath);
+            xpath.setf("Resource[@resourcePath='%s'][1]", respath.str());
+        }
+
+        return manifest->queryPropTree(xpath.str());
+    }
+
+    bool ends_with(const std::string &str, const std::string &suffix) const
+    {
+        if (suffix.size() > str.size())
+            return false;
+        return std::equal(suffix.rbegin(), suffix.rend(), str.rbegin());
+    }
+
+    //  --- IManifest ---
+    virtual const char *extractResources(StringBuffer &sb) const
+    {
+        // dll->queryManifestFiles extracts the manifest files to a tmp folder the paths as a StringArray
+        const StringArray &resFiles = dll->queryManifestFiles("UNKNOWN", "nlp");
```

Comment:
@afishbeck The use of "UNKNOWN" here, feels risky?

Reply:
I think any resource that isn't assigned a type in the manifest is given a type of unknown.  Type was primarily to distinguish resources ESP used directly like XSLT or HTML.  But I'd have to refresh my memory.  Perhaps queryManifestFiles should allow calling it without a type.  Odds are based on glancing at the code that dll->queryManifestFiles("*", "nlp"); would work but it would need to be tested.

Reply:
If your manifest file added these resources with the type='nlp' then you could use "nlp" to locate them here.

Reply:
I think that will be preferable, as there could be other "resources" included (say from the Visualizer bundle).

Reply:
Add ticket:  https://track.hpccsystems.com/browse/HPCC-29697